msg.go

```diff
@@ -1359,14 +1333,191 @@ func fromBase64(s []byte) (buf []byte, err error) {
+func packStructUint32(i uint32, msg []byte, off int, lenmsg int) (off1 int, err error) {
+	if off+4 > lenmsg {
+		return lenmsg, &Error{err: "overflow packing uint32"}
+	}
+	msg[off], msg[off+1], msg[off+2], msg[off+3] = packUint32(i)
+	return off + 4, nil
+}
+
+// Truncates the byte slice in msg to match the expected length of the RR.
+// Returns an error if the slice is smaller than the expected size.
+func truncateMsgFromRdlength(msg []byte, off int, rdlength uint16) (truncmsg []byte, err error) {
+	lenmsg := len(msg)
+	lenrd := off + int(rdlength)
+	if lenrd > lenmsg {
+		return msg, &Error{err: "overflowing header size"}
+	}
+	msg = msg[:lenrd]
+	return msg, nil
+}
+
+// Unpack an RR header, returning the offset to the end of the header and a
+// re-sliced msg according to the expected length of the RR.
+func unpackHeader(hdr *RR_Header, msg []byte, off int) (off1 int, truncmsg []byte, err error) {
+	lenmsg := len(msg)
+	if off == lenmsg {
+		return off, msg, nil
+	}
+	hdr.Name, off, err = UnpackDomainName(msg, off)
+	if err != nil {
+		return lenmsg, msg, err
+	}
+	hdr.Rrtype, off, err = unpackStructUint16(msg, off, lenmsg)
+	if err != nil {
+		return lenmsg, msg, err
+	}
+	hdr.Class, off, err = unpackStructUint16(msg, off, lenmsg)
+	if err != nil {
+		return lenmsg, msg, err
+	}
+	hdr.Ttl, off, err = unpackStructUint32(msg, off, lenmsg)
+	if err != nil {
+		return lenmsg, msg, err
+	}
+	hdr.Rdlength, off, err = unpackStructUint16(msg, off, lenmsg)
+	if err != nil {
+		return lenmsg, msg, err
+	}
+	msg, err = truncateMsgFromRdlength(msg, off, hdr.Rdlength)
+	return off, msg, nil
+}
+
+// Unpacks the IP stored in an A record.
+func unpackA(a *net.IP, msg []byte, off int) (off1 int, err error) {
+	lenmsg := len(msg)
+	if off == lenmsg {
+		return off, nil // dyn. update
+	}
+	if off+net.IPv4len > lenmsg {
+		return lenmsg, &Error{err: "overflow unpacking a"}
+	}
+	*a = net.IPv4(msg[off], msg[off+1], msg[off+2], msg[off+3])
+	off += net.IPv4len
+	return off, nil
+}
+
+// Pack an RR header, returning the offset to the end of the header.
+// See PackDomainName for documentation about the compression.
+func packHeader(hdr RR_Header, msg []byte, off int, compression map[string]int, compress bool) (off1 int, err error) {
+	lenmsg := len(msg)
+	if off == lenmsg {
+		return off, nil
+	}
+	off, err = PackDomainName(hdr.Name, msg, off, compression, compress)
+	if err != nil {
+		return lenmsg, err
+	}
+	off, err = packStructUint16(hdr.Rrtype, msg, off, lenmsg)
+	if err != nil {
+		return lenmsg, err
+	}
+	off, err = packStructUint16(hdr.Class, msg, off, lenmsg)
+	if err != nil {
+		return lenmsg, err
+	}
+	off, err = packStructUint32(hdr.Ttl, msg, off, lenmsg)
+	if err != nil {
+		return lenmsg, err
+	}
+	off, err = packStructUint16(hdr.Rdlength, msg, off, lenmsg)
+	if err != nil {
+		return lenmsg, err
+	}
+	return off, nil
+}
+
+// Packs the IP stored in an A record.
+func packA(a net.IP, msg []byte, off int) (off1 int, err error) {
```

Comment:
Longer term I feel go dns will be moving to a reflection-less impl. so merging this makes sense. Does it also make sense to make these pack functions part of the RR interface? So they become methods on the type?